### ISSUE TITLE
hclfmt command does not work as a pipeline

### ISSUE THREAD

**Issue by mbainter** [gruntwork-io/terragrunt#1037]
The `terraform fmt` command offers the ability to pipe a file into stdin and get the formatted result on stdout, which is used by various plugins to help with formatting files correctly in real-time within an editor. I'm working on adding in support for formatting .hcl files with terragrunt but I'm stuck on the problem that you can't use stdin/stdout in this way with terragrunt's hclfmt command.

Can support for this be added?

**Comment by yorinasub17** [comment]
We don't have a need for this use case internally at gruntwork, but if anyone wants to add a PR for it that would be great!

We actually use the terraform fixers in our editors. E.g here is my vim config to make this work:

```
au! BufNewFile,BufRead terragrunt.hcl set filetype=terraform syntax=terraform
```

`terraform fmt` works just fine with `terragrunt.hcl` in a pipeline because it can't do the regular filename check when reading from stdin. With that said, there is no guarantee it will continue to keep working, which was why `hclfmt` was added. So adding the pipeline support for when that breaks would be good to have.

**Comment by mbainter** [comment]
Yeah, I have something similar, but I wasn't certain that it would catch everything.  I assumed there was a reason you included an hclfmt instead of just recommending that route, was that incorrect?

**Comment by yorinasub17** [comment]
`hclfmt` was added because it works better in a CI pipeline, where you want to find, run, and update all the files from a root directory. In this use case, `terraform fmt` fails because it does filename checks.

The other reason is what I mentioned in the previous comment:

> With that said, there is no guarantee it will continue to keep working, which was why hclfmt was added

This is based on implication that the terraform team was going to start specializing formatting to be specific to terraform: https://github.com/hashicorp/terraform/pull/22092

**Comment by Mayeu** [comment]
@mbainter I found [rhadley-recurly/vim-terragrunt](https://github.com/rhadley-recurly/vim-terragrunt) today which support format on save for hcl files. Does that fill your need?

**Comment by CaptainPatate** [comment]
For the ones who want something better than `terragrunt hclfmt`, there is an `hclfmt` command available [in the official HCL repo](https://github.com/hashicorp/hcl/blob/v2.7.0/cmd/hclfmt/main.go) (but there is no prebuilt executable).

**Comment by yorinasub17** [comment]
FWIW, the `hclfmt` functionality uses the same functions as the official `hclfmt` version (https://github.com/gruntwork-io/terragrunt/blob/master/cli/hclfmt.go#L65), with the main difference being that it looks for and only runs on terragrunt config files.

**Comment by CaptainPatate** [comment]
@yorinasub17 yes I should have said "more flexible" instead of "better" because the formatter is indeed the same.

**Comment by caseycs** [comment]
In the meantime simple wrapper bash script can help:

```shell
#!/bin/sh
# wrapper to format hcl passed via stdin
# user for for VSCcode Custom Local Formatters extension (https://marketplace.visualstudio.com/items?itemName=jkillian.custom-local-formatters)
# or any other editor relying on stdin/stdout for formatting, like Zed
# relevant issues:
# * https://github.com/gruntwork-io/terragrunt/issues/1037
# * https://github.com/gruntwork-io/terragrunt/issues/2926
TMPFILE=$(mktemp)
cat - > $TMPFILE
terragrunt hclfmt --terragrunt-hclfmt-file $TMPFILE
cat $TMPFILE
rm $TMPFILE
```

**Comment by andrewdcato** [comment]
> @mbainter I found [rhadley-recurly/vim-terragrunt](https://github.com/rhadley-recurly/vim-terragrunt) today which support format on save for hcl files. Does that fill your need?

Working great for me in Neovim in 2025! Installed via Lazy: 

```lua
{
  "rhadley-recurly/vim-terragrunt",
  lazy = true,
  event = "BufEnter",
  config = function()
    vim.g.hcl_fmt_autosave = 0
  end,
}
```

**Comment by yhakbar** [comment]
Given that we've got support for formatting from stdin as of [v0.68.12](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.68.12) (thanks to @alikhil), I'm closing this issue out.

Please note that this flag will be renamed as part of the [CLI Redesign](https://github.com/gruntwork-io/terragrunt/issues/3445), so you'll need to stay tuned to make sure you aren't using a deprecated flag. We will have a moderately long period where the deprecated flag is supported prior to removing it.

In the future, you'll be able to use this functionality like so:

```bash
terragrunt hcl fmt --stdin
```
